### ISSUE TITLE
core: push SOCI index when publishing docker image assets

### ISSUE THREAD

**Issue by tmokmss** [aws/aws-cdk#26413]
### Describe the feature

Recently Seekable OCI (SOCI) is supported on ECS Fargate ([blog](https://aws.amazon.com/blogs/aws/aws-fargate-enables-faster-container-startup-using-seekable-oci/)).

> Amazon ECS tasks on Fargate that use Linux platform version 1.4.0 can use Seekable OCI (SOCI) to help start tasks faster. With SOCI, containers only spend a few seconds on the image pull before they can start, providing time for environment setup and application instantiation while the image is downloaded in the background. This is called lazy loading. When Fargate starts an Amazon ECS task, Fargate automatically detects if a SOCI index exists for an image in the task and starts the container without waiting for the entire image to be downloaded.
https://docs.aws.amazon.com/AmazonECS/latest/userguide/container-considerations.html#fargate-tasks-soci-images

It can help if CDK automatically builds and publishes a SOCI index for a docker image assets.

### Use Case

To easily publishes SOCI index from CDK, and eventually increasing the speed of Fargate cold start for large container images.

### Proposed Solution

Run [soci-snapshotter](https://github.com/awslabs/soci-snapshotter) locally on cdk deploy, and publish the artifact to ecr. Whether cdk publishes SOCI index or not should be optional because sometimes it will slow down fargate startup time especially for small images ([ref](https://twitter.com/apparentorder/status/1681321222225928193)). 

### Other Information

There is already a solution to automatically build and push SOCI index when an image is pushed to ECR.

https://aws-ia.github.io/cfn-ecr-aws-soci-index-builder/

Because this solution publishes a SOCI index asynchronously, sometimes the index is not ready yet when a Fargate task launches. The CDK solution might be better in that we can easily set dependencies between assets and ECS task definition, which allows to wait an update of task definition until the corresponding SOCI index is available.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.87.0

### Environment details (OS name and version, etc.)

macOS

**Comment by khushail** [comment]
Thanks for submitting this feature request @tmokmss.

**Comment by pahud** [comment]
Thank you @tmokmss 

Where would you expect the index to be built? From the local environment where CDK bundle and publish the assets or from the lambda function? I guess the best solution might be transparently build the index when CDK upload the container image assets but this will require the client to install the `soci-snapshotter`. If users opt in the index building and the `soci-snapshotter` is installed, maybe the image bundling process can build and upload the index transparently?

**Comment by tmokmss** [comment]
Personally building it locally is preferable, because using CFn custom resource would introduce some overhead such as triggering Lambda from CFn, pulling images from ECR, etc. It would also have several limitations like RAM or 15min duration (idk how long it usually takes to create an index though.)

That said, soci-snapshotter [appears](https://github.com/awslabs/soci-snapshotter/blob/main/docs/getting-started.md#install-soci-snapshotter) to only run on Linux to build an index (I'm confirming it on [this thread](https://github.com/awslabs/soci-snapshotter/discussions/733)). If so, I guess we have to use a custom resource after all.

**Comment by tmokmss** [comment]
For anyone interested in this issue, I published an experimental construct to build and publish a SOCI index: [deploy-time-build](https://github.com/tmokmss/deploy-time-build).

You can push a SOCI index during CFn deployment by the following code: 

```ts
import { SociIndexBuild } from 'deploy-time-build;

const asset = new DockerImageAsset(this, 'Image', { directory: 'example-image' });
SociIndexBuild.fromImageAsset(this, 'Index', asset);
```

**Comment by pahud** [comment]
Hi @tmokmss  does it mean the client will need to install additional tool for that?

**Comment by tmokmss** [comment]
@pahud No, the above construct works as a custom resource on Lambda and CodeBuild, so there's no dependency on a local machine. As far as I researched, building a SOCI index locally is difficult for now.

**Comment by BwL1289** [comment]
Interested in this. Would love to see a CDK python solution.

**Comment by tmokmss** [comment]
@BwL1289 You can use deploy-time-build in Python as well.

Install via pip:

```
pip install deploy-time-build
```

then

```py
from deploy_time_build import SociIndexBuild

asset = DockerImageAsset(self, "Image", directory="example-image")
SociIndexBuild.from_docker_image_asset(self, "Index", asset)
```

Here's [the doc](https://constructs.dev/packages/deploy-time-build/v/0.3.24?lang=python#build-soci-index-for-a-container-image) for more details.

**Comment by BwL1289** [comment]
Ack! Thank you!

**Comment by BwL1289** [comment]
@tmokmss question I couldn't find an answer to looking at the docs: 

while the construct will create the soci index for me, I still need to implement the soci-snapshotter myself right?

**Comment by tmokmss** [comment]
@BwL1289 As long as you are using ECS Fargate, the soci index is automatically utilized and you do not need to care about image loading part ([related blog](https://aws.amazon.com/blogs/containers/under-the-hood-lazy-loading-container-images-with-seekable-oci-and-aws-fargate/).)

**Comment by BwL1289** [comment]
Ack thank you. Looks like for other services (i.e. CodeBuild) the snapshotter may be needed.

FWIW - Was able to successfully use SociIndex to build soci indexes. Appreciate your hard work.
